![image](https://wallpaperaccess.com/full/1670957.jpg)

# Learning How to Clean Data in Python - Analyzing Car Classifieds on EBay

The objective of this project is to provide an example of how to clean data with Python.

We utilized a used-cars dataset found in a classifieds section of the German eBay website. The original dataset was scraped fom eBay Kleinanzeigan and is available on [kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). For this analysis, we used a sample of the original dataset with 50,000 that was prepared by [Dataquest](https://www.dataquest.io/) with a less-cleaned version of the data.

## Load and explore the dataset

In [1]:
# read dataset 
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding='latin-1')

In [2]:
# explore dataset
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

The dataset includes 50,000 entries and 20 columns. Five of the columns have missing values, since the number of non-null rows they have is lower than the total number of entries. However, no column has more than 20% of missing values, what shows that all might have enough data for analysis.

Most columns are of `object` type, these ones are probably of `str`type, while 5 are of `int` type.

Let's have a look at the first five rows to have more insight on the data.

In [3]:
# show first five rows of the dataset
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can see that the first column `dateCrowled` includes `datetime` data. The `name` column includes both text and numeric data. The `price` column includes a dollar sign, what prevent us from making calculations. Still, column names are of `camelcase` instead of `snakecase`, which would be the usual format. 

## Renaming Column Labels

In [4]:
# print the column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
# adjust column names
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

# replace columns by new columns
autos.columns = new_columns

In [6]:
# print columns with new names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

In [7]:
# display first five columns with the revised labels
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

Now, column names are more informative, easier to understand and analyze. Next, lets explore further the data to identify the need for cleaning.

## Display Descriptive Statistics

In [8]:
# print descriptive statistics for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Analyzing the results above we see that:
- `seller`, `offer_type` and `num_photos` seem to have only one value for all rows. These columns should be dropped.
- Columns to be further investigated: `name`, `registration_year`, `registration_month` and `ad_created`.
- Numeric data to be cleaned: `price` and `odometer`.

## Drop Irrelevant Columns
We will drop the `seller`, `offer_type` and `num_photos` columns.

In [9]:
# drop columns seller, offer_type and num_photos
autos.drop(['seller', 'offer_type', 'num_photos'], axis='columns', inplace=True)
print(autos.shape)
print(autos.columns)

(50000, 17)
Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


As we can see above, the three columns are not anymore part of the dataset. We can also observe that the total number of columns is 17 instead of 20.

## Analyze Suspicious Columns

In [10]:
# investigate the name column
autos['name'].value_counts(dropna=False).head(16)

Ford_Fiesta            78
Volkswagen_Golf_1.4    75
BMW_316i               75
BMW_318i               72
Volkswagen_Polo        72
BMW_320i               71
Opel_Corsa             71
Renault_Twingo         70
Volkswagen_Golf        57
Opel_Corsa_1.2_16V     56
BMW_116i               53
Peugeot_206            52
Opel_Corsa_B           52
Ford_Focus             50
Volkswagen_Polo_1.2    48
Volkswagen_Golf_1.6    47
Name: name, dtype: int64

We see that the `name`column is very granular, differentiating specific versions of the same models (e.g., Volkswagen_Golf_1.4 and Volkswagen_Golf_1.6). Let's have a look at the `registration_month` column.

In [11]:
# investigate the registration_month column
autos['registration_month'].value_counts(dropna=False)

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

We see that there are 5,075 entries with values `0` that cannot be months. These will need to be considered missing.

## Dealing with Missing Values

Part of the values in the `registration_month` column are missing. Let's classify them accordingly.

In [12]:
# replacing '0' values in the registration_month column as missing
autos.loc[autos['registration_month']==0, 'registration_month'] = np.nan

In [13]:
autos.registration_month.value_counts(dropna=False)

NaN     5075
3.0     5071
6.0     4368
5.0     4107
4.0     4102
7.0     3949
10.0    3651
12.0    3447
9.0     3389
11.0    3360
1.0     3282
8.0     3191
2.0     3008
Name: registration_month, dtype: int64

We see above that the values are now displayed as missing. We continue the analysis investigating the `ad_created` column.

In [14]:
# investigate the ad_created column
autos['ad_created'].value_counts(dropna=False).head()

2016-04-03 00:00:00    1946
2016-03-20 00:00:00    1893
2016-03-21 00:00:00    1886
2016-04-04 00:00:00    1844
2016-03-12 00:00:00    1831
Name: ad_created, dtype: int64

This columns looks just fine. 

## Cleaning the `price` and `odometer` columns
They include non-numeric characters as dollar signs and commas. Let's remove these characters and convert these columns to numeric.

In [15]:
# clean the price column
autos['price'] = autos['price'].str.replace('$', '') # removing dollar signs
autos['price'] = autos['price'].str.replace(',', '') # removing commas 
autos['price'] = autos['price'].astype(float) # converting data type into float
print(autos['price'].dtype)
print(autos['price'].head())

float64
0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64


In [16]:
# clean the odometer column
autos['odometer'] = autos['odometer'].str.replace('km', '') # remove km
autos['odometer'] = autos['odometer'].str.replace(',', '') # remove commas
autos['odometer'] = autos['odometer'].astype(int) # convert column into integer type
print(autos['odometer'].dtype)
print(autos['odometer'].head())

int64
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64


## Renaming the Odometer Column
Then, we rename the `odometer` column to `odometer_km`.

In [17]:
# rename the odometer column to odometer_km
autos.rename({'odometer':'odometer_km'}, axis='columns', inplace=True)
print(autos.columns)

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


## Explore the Numeric Columns
We start investigating extreme values.

In [18]:
# Display minimum and maximum values
print('Minimum Price:', autos.price.min())
print('Maximum Price:', autos.price.max())
print('Minimum Mileage:', autos.odometer_km.min())
print('Maximum Mileage:', autos.odometer_km.max())

Minimum Price: 0.0
Maximum Price: 99999999.0
Minimum Mileage: 5000
Maximum Mileage: 150000


The `odometer_km`column seems to make sense, however the minumum and maximum prices are too extreme and should be removed. In the following lines we will continue to explore these two columns.

In [19]:
# print the naumber of unique values for odometer and price columns
print('Number of unique prices: ', autos.odometer_km.unique().shape[0]) 
print('Number of unique odometer values: ', autos.price.unique().shape[0])

Number of unique prices:  13
Number of unique odometer values:  2357


In [20]:
# display descriptive statistics of these two columns
print(autos.odometer_km.describe())
print('\n')
print(autos.price.describe())

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [21]:
# display value counts for the odometer column
autos.odometer_km.value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [22]:
#display the highest price values
autos.price.value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

In [23]:
# display the lowest price values
autos.price.value_counts().sort_index(ascending=True).head(15)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
Name: price, dtype: int64

From this analysis we see that the `odometer_km` column seems ok. However, the price column includes values that seem unrealistically high and low. There are 1,421 rows with price iquals zero. On the other hand, there are very high values above the threshold of 350,000.

# Drop Unrealistic Prices
We will drop all rows with prices below 100 or above 350,000 dollars.

In [24]:
# drop all rows with price values below 100 or above 350,000 dollars.
autos = autos[autos['price'].between(100, 350000)]

In [25]:
# display descriptive statistics for the price column
autos.price.describe()

count     48224.000000
mean       5930.371433
std        9078.372762
min         100.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

The new dataset after removing the extreme prices still has more than 48,224 rows. We were able to improve the quality of the data removing less than 4% of the data.

## Explore Datetime Columns
Next, we will analyze the values of three columns: `date_crawled`, `ad_created`, `last_seen`.

In [26]:
# analize the values in the column date_crawled
(autos.date_crawled
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.025361
2016-03-06    0.014039
2016-03-07    0.036061
2016-03-08    0.033179
2016-03-09    0.033013
2016-03-10    0.032287
2016-03-11    0.032598
2016-03-12    0.036911
2016-03-13    0.015677
2016-03-14    0.036662
2016-03-15    0.034319
2016-03-16    0.029467
2016-03-17    0.031499
2016-03-18    0.012898
2016-03-19    0.034734
2016-03-20    0.037803
2016-03-21    0.037201
2016-03-22    0.032888
2016-03-23    0.032287
2016-03-24    0.029446
2016-03-25    0.031499
2016-03-26    0.032308
2016-03-27    0.031126
2016-03-28    0.034962
2016-03-29    0.034112
2016-03-30    0.033738
2016-03-31    0.031851
2016-04-01    0.033697
2016-04-02    0.035605
2016-04-03    0.038611
2016-04-04    0.036538
2016-04-05    0.013064
2016-04-06    0.003173
2016-04-07    0.001389
Name: date_crawled, dtype: float64

The `date_crawled`column includes values from March and April 2016.

In [27]:
# analize values in the ad_created column
(autos.ad_created
.str[:10]
.value_counts(normalize=True, dropna=False)
.sort_index()
)             

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033096
2016-03-10    0.031997
2016-03-11    0.032909
2016-03-12    0.036745
2016-03-13    0.017045
2016-03-14    0.035294
2016-03-15    0.034049
2016-03-16    0.029964
2016-03-17    0.031167
2016-03-18    0.013582
2016-03-19    0.033614
2016-03-20    0.037865
2016-03-21 

The `ad_created` column includes values from 2015 and 2016. 

In [28]:
# analize values in the last_seen column
(autos.last_seen
.str[:10]
.value_counts(normalize=True, dropna=False)
.sort_index()
)

2016-03-05    0.001078
2016-03-06    0.004313
2016-03-07    0.005433
2016-03-08    0.007320
2016-03-09    0.009580
2016-03-10    0.010638
2016-03-11    0.012400
2016-03-12    0.023785
2016-03-13    0.008875
2016-03-14    0.012629
2016-03-15    0.015863
2016-03-16    0.016444
2016-03-17    0.028098
2016-03-18    0.007320
2016-03-19    0.015760
2016-03-20    0.020654
2016-03-21    0.020550
2016-03-22    0.021359
2016-03-23    0.018580
2016-03-24    0.019762
2016-03-25    0.019098
2016-03-26    0.016672
2016-03-27    0.015552
2016-03-28    0.020840
2016-03-29    0.022292
2016-03-30    0.024697
2016-03-31    0.023826
2016-04-01    0.022852
2016-04-02    0.024884
2016-04-03    0.025133
2016-04-04    0.024531
2016-04-05    0.125062
2016-04-06    0.221964
2016-04-07    0.132154
Name: last_seen, dtype: float64

The `last_seen` column includes values from 2016. Let's analyze the `registration_year` column.

## Explore Registration Year

In [29]:
autos['registration_year'].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The `registration_year` column includes strange values. The minimum is the year 1000 and the maximum the year 9999. 

In [30]:
print(autos['registration_year'].shape)
print(autos['registration_year'].value_counts().sort_index())

(48224,)
1000       1
1001       1
1111       1
1800       2
1910       2
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      22
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3104
2001    2629
2002    2477
2003    2693
2004    2699
2005    2911
2006    2668
2007    2273
2008    2210
2009    2080
2010    1587
2011    1618
2012    1308
2013     801
2014     662
2015     380
2016    1202
2017    1383
2018     468
2019       2
2800       1
4100       1
4500       1
4800       1
5000       3
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64


From the results above, se see that before 1910 the dates get back to 1800. On the other hand,  the columns `date_crawled` and `last_seen` include dates that go until 2016 to the latest.

We will consider only dates that range from 1910 to 2016. Next, we will remove the rows outside this range.

In [31]:
# calculate the percentage of rows with `registration_year` values outside the range expected
(~autos['registration_year'].between(1910, 2016)).sum() / autos['registration_year'].shape[0]

0.03881884538818845

As we see above, less than 4% of the `registration_year` values are out of a reasonable range. 
Next, we will remove the values out of the range.

## Drop Rows with Registration Year out of Range
We will include only entries with `registration_year` between 1910 and 2016.

In [32]:
# Remove entries outside registration_year reasonable range
autos = autos[autos['registration_year'].between(1910, 2016)]

In [33]:
print('Number of rows: ', autos['registration_year'].shape[0])
print(autos['registration_year'].value_counts(normalize=True).sort_index())

Number of rows:  46352
1910    0.000043
1927    0.000022
1929    0.000022
1931    0.000022
1934    0.000043
1937    0.000086
1938    0.000022
1939    0.000022
1941    0.000043
1943    0.000022
1948    0.000022
1950    0.000022
1951    0.000043
1952    0.000022
1953    0.000022
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000475
1961    0.000129
1962    0.000086
1963    0.000173
1964    0.000259
1965    0.000367
1966    0.000475
1967    0.000561
1968    0.000561
          ...   
1987    0.001553
1988    0.002869
1989    0.003689
1990    0.007163
1991    0.007292
1992    0.007918
1993    0.009061
1994    0.013505
1995    0.025738
1996    0.029233
1997    0.041530
1998    0.050483
1999    0.062112
2000    0.066966
2001    0.056718
2002    0.053439
2003    0.058099
2004    0.058228
2005    0.062802
2006    0.057560
2007    0.049038
2008    0.047679
2009    0.044874
2010    0.034238
2011    0.034907
2012    0.028219
2013    

After the cleaning, the registration_year includes 46,352 values ranging from 1910 to 2016. Next, we will calculate the mean price by brand.

## Analyze Brands' Average Prices and Mileage
Next, we will identify the Top10 car brands and calculate their average price and mileage.

In [34]:
# display unique values for the brand column
autos.brand.unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [35]:
# show the top 10 brands in terms of the percentage of the entries
top10_brands = autos.brand.value_counts(normalize=True).sort_values(ascending=False).head(10)
top10_index = top10_brands.index
print(top10_index)
print(top10_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')
volkswagen       0.211404
bmw              0.110179
opel             0.107245
mercedes_benz    0.096652
audi             0.086771
ford             0.069835
renault          0.047075
peugeot          0.029858
fiat             0.025608
seat             0.018252
Name: brand, dtype: float64


We will focus on the Top10 brands, the ones more frequently found in the website. In the following lines, we will calculate the average price and milleage for each of the Top10 brands.

### Calculate Average Price and Mileage for the Top10 Brands

In [36]:
# create two dictionaries, one with average prices and another with average milleage for the top10 brands
avg_price = {}
avg_mileage = {}

for car in top10_index:
    col_price = autos.loc[autos['brand']==car, 'price']
    mean_price = round(col_price.mean())
    col_mileage = autos.loc[autos['brand']==car, 'odometer_km']
    mean_mileage = round(col_mileage.mean())
    avg_price[car]=mean_price
    avg_mileage[car]=mean_mileage
    
print('Mean Prices Dictionary:')
print(avg_price)
print('\n')
print('Mean Mileage Dictionary:')
print(avg_mileage)

Mean Prices Dictionary:
{'volkswagen': 5437, 'bmw': 8382, 'opel': 3005, 'mercedes_benz': 8673, 'audi': 9381, 'ford': 3779, 'renault': 2496, 'peugeot': 3114, 'fiat': 2837, 'seat': 4433}


Mean Mileage Dictionary:
{'volkswagen': 128800, 'bmw': 132695, 'opel': 129384, 'mercedes_benz': 131026, 'audi': 129245, 'ford': 124277, 'renault': 128281, 'peugeot': 127128, 'fiat': 116950, 'seat': 121537}


### Convert Dictionaries into Series and Dataframes

In [37]:
# convert both dictionaries into series
top10_price_series = pd.Series(avg_price)
top10_mileage_series = pd.Series(avg_mileage)
print(top10_price_series)
print('\n')
print(top10_mileage_series)

volkswagen       5437
bmw              8382
opel             3005
mercedes_benz    8673
audi             9381
ford             3779
renault          2496
peugeot          3114
fiat             2837
seat             4433
dtype: int64


volkswagen       128800
bmw              132695
opel             129384
mercedes_benz    131026
audi             129245
ford             124277
renault          128281
peugeot          127128
fiat             116950
seat             121537
dtype: int64


In [38]:
# convert the series top10_price_series into a DataFrame
top10_df = pd.DataFrame(top10_price_series, columns=['mean_price'])

In [39]:
print(top10_df)

               mean_price
volkswagen           5437
bmw                  8382
opel                 3005
mercedes_benz        8673
audi                 9381
ford                 3779
renault              2496
peugeot              3114
fiat                 2837
seat                 4433


In [40]:
# assign the milleage series into a column of the price dataframe
top10_df['mean_mileage']=top10_mileage_series
print(top10_df.sort_values('mean_price', ascending=False))

               mean_price  mean_mileage
audi                 9381        129245
mercedes_benz        8673        131026
bmw                  8382        132695
volkswagen           5437        128800
seat                 4433        121537
ford                 3779        124277
peugeot              3114        127128
opel                 3005        129384
fiat                 2837        116950
renault              2496        128281


Above we compare the Top10 brands in the dataset in terms of their average prices and mileage. We have no evidence of a relationaship between the two columns, since the highest and lowest prices present similar average mileage values.

## Translating German Names into English

We will identify German names in the datafile and translate them into English. We see German therms in the columns `gearbox` and `unrepaired_damage`.

In [41]:
# display values in the gearbox column
print(autos.gearbox.value_counts(normalize=True).sort_values(ascending=False))
print('\n')
print(autos.unrepaired_damage.value_counts(normalize=True).sort_values(ascending=False))

manuell      0.778471
automatik    0.221529
Name: gearbox, dtype: float64


nein    0.882836
ja      0.117164
Name: unrepaired_damage, dtype: float64


Let's translate the word 'manuell' as 'manual' and 'automatik' into 'automatic'. We will also translate 'nein' into 'no' and 'ja' into 'yes'.

In [42]:
# replace terms in German in the gearbox column
autos.gearbox = autos.gearbox.str.replace('manuell', 'manual')
autos.gearbox = autos.gearbox.str.replace('automatik', 'automatic')
print(autos.gearbox.value_counts(normalize=True).sort_values(ascending=False))
print('\n')
autos.unrepaired_damage = autos.unrepaired_damage.str.replace('nein', 'no')
autos.unrepaired_damage = autos.unrepaired_damage.str.replace('ja', 'yes')
print(autos.unrepaired_damage.value_counts(normalize=True).sort_values(ascending=False))

manual       0.778471
automatic    0.221529
Name: gearbox, dtype: float64


no     0.882836
yes    0.117164
Name: unrepaired_damage, dtype: float64


We can confirm above that both the `gearbox` and `unrepaired_damage` columns values are not displayed in English.

## Extracting the Date from Datetime Values
The column `date_crawled` includes values that are not appropriate for analysis. In the code below we show first how the data is displayed in the dataset. 

In [43]:
# display the first 5 rows of the date_crawled values
autos.date_crawled.value_counts(normalize=True).sort_index().head(5)

2016-03-05 14:06:30    0.000022
2016-03-05 14:06:40    0.000022
2016-03-05 14:07:04    0.000022
2016-03-05 14:07:21    0.000022
2016-03-05 14:07:26    0.000022
Name: date_crawled, dtype: float64

Next we will create a new column showing only the date numeric values, which includes the 10 first digits from each entry.

In [44]:
# create a new column date_crawled_numeric including only the date with numeric values
autos['date_crawled_numeric'] = (autos.date_crawled
                              .str[:10]
                              .str.replace('-', '')
                              .astype(int)
                             )
print(autos.date_crawled_numeric.value_counts(normalize=True).sort_index().head(5))
print(autos.date_crawled_numeric.dtype)

20160305    0.025220
20160306    0.014153
20160307    0.036331
20160308    0.033418
20160309    0.033181
Name: date_crawled_numeric, dtype: float64
int64


## Finding Out the Most Common Brand/Model Combinations
Brand and Model are displayed in two different columns. We will combine these two columns into a new one called `brand_model`.

In [45]:
# create a new column brand_model
autos['brand_model'] = autos.brand + '_' + autos.model

In [46]:
autos.brand_model.value_counts(normalize=True).head(5)

volkswagen_golf      0.083309
bmw_3er              0.058841
volkswagen_polo      0.036001
opel_corsa           0.035458
volkswagen_passat    0.030415
Name: brand_model, dtype: float64

We see that Volkswagen Gold and BMW 3er are the most common combinations. Next, we will split odometer_km into groups and check the average prices per group.

## Create Groups According to Mileage and Calculate Average Prices per Group

In [47]:
# split odometer_km in groups and calculate average prices per group
#first check the range
autos.odometer_km.describe()

count     46352.000000
mean     125734.488264
std       39608.848152
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

As we see above, it ranges from 5,000km to 150,000km. Let's split the DataFrame in 4 groups:
- Group1: Between 5,000 and 49,999
- Group2: Between 50,000 and 74,999
- Group3: Between 75,000 and 99,999
- Group4: Between 100,000 and 150,000

In [48]:
# Split the dataset in 4 groups
group1 = autos[autos['odometer_km'].between(5000, 49999)]
group2 = autos[autos['odometer_km'].between(50000, 74999)]
group3 = autos[autos['odometer_km'].between(75000, 99999)]
group4 = autos[autos['odometer_km'].between(100000, 150000)]

Next, we calculate the average price for each group.

In [49]:
# create an empty list with the four groups and 
#an empty dictionary for the average prices per group
groups_list = [group1, group2, group3, group4]
avg_price_per_group = {} 
group_name = 0

# loop over the group list and append the average prices to the empty dictionary
for group in groups_list:
    price_series = group['price']
    price_mean = round(price_series.mean())
    group_name += 1
    group_name_str = str(group_name)
    group_name_final = 'Group ' + group_name_str
    avg_price_per_group[group_name_final] = price_mean
    
print(avg_price_per_group)

{'Group 1': 15631, 'Group 2': 12320, 'Group 3': 9044, 'Group 4': 4358}


As we can observe above, the average price is lower for higher mileage.
As a final exercize, lets compare the average price between cars with and without damages.

## Compare Average Prices for Cars With and Without Damages

In [50]:
# expore the unrepaired_damage column
autos.unrepaired_damage.value_counts(dropna=False)

no     33772
NaN     8098
yes     4482
Name: unrepaired_damage, dtype: int64

In [51]:
# calculate the average price for cars with and withoug damage
damage_list = ['yes', 'no']
avg_damage = {}
for damage in damage_list:
    damage_df = autos[autos['unrepaired_damage']==damage]
    damage_series = damage_df['price']
    damage_mean = round(damage_series.mean())
    avg_damage[damage] = damage_mean
    
print(avg_damage)
    

{'yes': 2270, 'no': 7177}


We can see that cars with some sort of damage have considerably lower prices if compared to the ones with damages.

## Conclusion
The aim of this project was to present how to clean data in Python. We used a dataset with data with car classifieds from eBay with 50,000 entries.

The project shows how to: (1) read a datafile; (2) explore the data; (3) drop irrelevant columns; (4) deal with missing values; (5) calculate descriptive statistics; (6) group data by specific columns and conduct calculations for each groups.